resources/images/model/images.go

```diff
@@ -45,42 +50,102 @@ func NewImagesModel(
 	}
 }
 
+// CreateImage parses artifact and uploads artifact file to the file storage - in parallel,
+// and creates image structure in the system.
+// Returns image ID and nil on success.
 func (i *ImagesModel) CreateImage(
-	artifact io.Reader,
 	metaConstructor *images.SoftwareImageMetaConstructor,
-	metaArtifactConstructor *images.SoftwareImageMetaArtifactConstructor) (string, error) {
-
-	if metaConstructor == nil || metaArtifactConstructor == nil {
+	imageReader io.Reader) (string, error) {
+	if metaConstructor == nil {
 		return "", controller.ErrModelMissingInputMetadata
 	}
+	if imageReader == nil {
+		return "", controller.ErrModelMissingInputArtifact
+	}
+	imageID, artifactFileID, err := i.handleArtifact(metaConstructor, imageReader)
+	// try to remove artifact file from file storage on error
+	if err != nil {
+		i.fileStorage.Delete(artifactFileID)
+	}
+	return imageID, err
+}
+
+// handleArtifact parses artifact and uploads artifact file to the file storage - in parallel,
+// and creates image structure in the system.
+// Returns image ID, artifact file ID and nil on success.
+func (i *ImagesModel) handleArtifact(
+	metaConstructor *images.SoftwareImageMetaConstructor,
+	imageReader io.Reader) (string, string, error) {
+
+	// limit just for safety
+	// max image size - 10G
+	const MaxImageSize = 1024 * 1024 * 1024 * 10
+
+	// create pipe
+	pR, pW := io.Pipe()
+	// limit reader to max image size
+	lr := io.LimitReader(imageReader, MaxImageSize)
+	tee := io.TeeReader(lr, pW)
+
+	artifactFileID := uuid.NewV4().String()
+
+	ch := make(chan error)
+	// create goroutine for artifact upload
+	go func() {
+		err := i.fileStorage.UploadArtifact(artifactFileID, pR, ImageContentType)
+		ch <- err
+	}()
+
+	// parse artifact
+	metaArtifactConstructor, err := getMetaFromArchive(&tee, MaxImageSize)
+	if err != nil {
+		pW.Close()
+		_ = <-ch
+		return "", "", controller.ErrModelInvalidMetadata
+	}
+
+	// read the rest of the data, just for safety
+	_, err = io.Copy(ioutil.Discard, tee)
+	if err != nil {
+		pW.Close()
```

Comment:
I'm not sure if there will be a profit from such solution, since we are closing PipeWriter in the middle of this method.